File: Kalorie/Kalorie/Core/UseCases/SearchFoodExternallyUseCase.swift
//
//  SearchFoodExternallyUseCase.swift
//  Kalorie
//
//  Created by Josef Antoni on 22.07.2026.
//

import Foundation

enum SearchFoodExternallyError: Error {
    case invalidURL
    case serverError(statusCode: Int)
}

protocol SearchFoodExternallyUseCaseProtocol {
    func callAsFunction(query: String) async throws -> [FoodItemDomain]
}

struct SearchFoodExternallyUseCase: SearchFoodExternallyUseCaseProtocol {

    // MARK: - Properties

    private let session: URLSession
    private let retryDelay: Duration

    // MARK: - Init

    init(session: URLSession = .shared, retryDelay: Duration = Constants.OpenFoodFacts.retryDelay) {
        self.session = session
        self.retryDelay = retryDelay
    }

    // MARK: - Functions

    func callAsFunction(query: String) async throws -> [FoodItemDomain] {
        var components = URLComponents(string: "https://\(Constants.OpenFoodFacts.host)/cgi/search.pl")
        components?.queryItems = [
            URLQueryItem(name: "search_terms", value: query),
            URLQueryItem(name: "json", value: "1"),
            URLQueryItem(name: "page_size", value: "20"),
            URLQueryItem(name: "fields", value: "code,product_name,product_name_cs,product_name_en,nutriments")
        ]
        guard let url = components?.url else { throw SearchFoodExternallyError.invalidURL }
        var request = URLRequest(url: url, timeoutInterval: Constants.OpenFoodFacts.requestTimeout)
        request.setValue(Constants.OpenFoodFacts.userAgent, forHTTPHeaderField: "User-Agent")
        let (data, statusCode) = try await OpenFoodFactsTransientRequest.data(for: request, session: session, retryDelay: retryDelay)
        guard (200...299).contains(statusCode) else {
            throw SearchFoodExternallyError.serverError(statusCode: statusCode)
        }
        let decoded = try JSONDecoder().decode(OpenFoodFactsResponseDTO.self, from: data)
        return decoded.products.compactMap { $0.asDomain() }
    }

}

#if DEBUG
struct SearchFoodExternallyUseCaseFake: SearchFoodExternallyUseCaseProtocol {

    // MARK: - Properties

    var shouldThrow = false
    var stubbedItems: [FoodItemDomain] = []

    // MARK: - Functions

    func callAsFunction(query: String) async throws -> [FoodItemDomain] {
        if shouldThrow { throw SearchFoodExternallyError.invalidURL }
        return stubbedItems
    }
}
#endif

File: Android/app/src/main/java/antoni/kalorie/core/usecases/SearchFoodExternallyUseCase.kt
package antoni.kalorie.core.usecases

import antoni.kalorie.core.models.FoodItemDomain
import antoni.kalorie.core.networking.HttpRequest
import antoni.kalorie.core.networking.HttpSessionProtocol
import antoni.kalorie.core.networking.HttpURLConnectionSession
import antoni.kalorie.core.networking.OpenFoodFactsResponseDTO
import antoni.kalorie.core.networking.OpenFoodFactsTransientRequest
import antoni.kalorie.core.utils.Constants
import java.net.URLEncoder
import kotlin.time.Duration
import kotlinx.serialization.json.Json

sealed class SearchFoodExternallyError : Exception() {
    data object InvalidURL : SearchFoodExternallyError()
    data class ServerError(val statusCode: Int) : SearchFoodExternallyError()
}

interface SearchFoodExternallyUseCaseProtocol {
    suspend operator fun invoke(query: String): List<FoodItemDomain>
}

class SearchFoodExternallyUseCase(
    private val session: HttpSessionProtocol = HttpURLConnectionSession(),
    private val retryDelay: Duration = Constants.OpenFoodFacts.RETRY_DELAY,
) : SearchFoodExternallyUseCaseProtocol {

    // MARK: - Functions

    override suspend fun invoke(query: String): List<FoodItemDomain> {
        val url = "https://${Constants.OpenFoodFacts.HOST}/cgi/search.pl" +
            "?search_terms=${URLEncoder.encode(query, Charsets.UTF_8).replace("+", "%20")}" +
            "&json=1&page_size=20&fields=$FIELDS"
        val request = HttpRequest(
            url = url,
            userAgent = Constants.OpenFoodFacts.USER_AGENT,
            timeoutMillis = Constants.OpenFoodFacts.REQUEST_TIMEOUT_MILLIS,
        )
        val response = OpenFoodFactsTransientRequest.data(request, session, retryDelay)
        if (response.statusCode !in 200..299) throw SearchFoodExternallyError.ServerError(response.statusCode)
        val decoded = json.decodeFromString<OpenFoodFactsResponseDTO>(response.body)
        return decoded.products.mapNotNull { it.asDomain() }
    }

    private companion object {
        const val FIELDS = "code,product_name,product_name_cs,product_name_en,nutriments"
        val json = Json { ignoreUnknownKeys = true }
    }
}
